src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.AnalyzerExecutionContext.cs

```diff
@@ -237,34 +237,43 @@ public void ClearSymbolScopeTask(ISymbol symbol)
             }
 
             public ImmutableArray<DiagnosticDescriptor> GetOrComputeDiagnosticDescriptors(DiagnosticAnalyzer analyzer, AnalyzerExecutor analyzerExecutor)
-                => GetOrComputeDescriptors(ref _lazyDiagnosticDescriptors, ComputeDiagnosticDescriptors, analyzer, analyzerExecutor);
+                => GetOrComputeDescriptors(ref _lazyDiagnosticDescriptors, ComputeDiagnosticDescriptors_NoLock, analyzer, analyzerExecutor, _gate);
 
             public ImmutableArray<SuppressionDescriptor> GetOrComputeSuppressionDescriptors(DiagnosticSuppressor suppressor, AnalyzerExecutor analyzerExecutor)
-                => GetOrComputeDescriptors(ref _lazySuppressionDescriptors, ComputeSuppressionDescriptors, suppressor, analyzerExecutor);
+                => GetOrComputeDescriptors(ref _lazySuppressionDescriptors, ComputeSuppressionDescriptors_NoLock, suppressor, analyzerExecutor, _gate);
 
             private static ImmutableArray<TDescriptor> GetOrComputeDescriptors<TDescriptor>(
                 ref ImmutableArray<TDescriptor> lazyDescriptors,
-                Func<DiagnosticAnalyzer, AnalyzerExecutor, ImmutableArray<TDescriptor>> computeDescriptors,
+                Func<DiagnosticAnalyzer, AnalyzerExecutor, ImmutableArray<TDescriptor>> computeDescriptorsNoLock,
                 DiagnosticAnalyzer analyzer,
-                AnalyzerExecutor analyzerExecutor)
+                AnalyzerExecutor analyzerExecutor,
+                object gate)
             {
                 if (!lazyDescriptors.IsDefault)
                 {
                     return lazyDescriptors;
                 }
 
-                // Otherwise, compute the value.
-                // We do so outside the lock statement as we are calling into user code, which may be a long running operation.
-                var descriptors = computeDescriptors(analyzer, analyzerExecutor);
+                lock (gate)
```

Comment:
isn't there an ImmutableArrayInterlocked helper for thsi?

Reply:
> isn't there an ImmutableArrayInterlocked helper for thsi pattern?

How does that help avoid multiple calls to `computeDescriptors` delegate?

Reply:
Also note that all of other callbacks into the analyzers are under a lock: https://github.com/dotnet/roslyn/blob/dc04e499a090e9e2277ef358eb385905fd4a6003/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs#L1603-L1626

We assumed it would be fine if `.SupportedDiagnostics` is invoked multiple times as it is not expensive and not expected to cause any semantic issues for the analyzer, but the fact that it can throw and lead to AD0001 means we cannot really allow duplicate callbacks.

Reply:
> How does that help avoid multiple calls to computeDescriptors delegate?

You call the version that takes a func and only invokes it if the value is not already set. (note: i'm not pushing you to use this.  just throwing this out as an option :)).

Reply:
> You call the version that takes a func and only invokes it if the value is not already set

I could not find a version of `ImmutableInterlocked.InterlockedInitialize` that takes a func. Is there any other helper that I can use?